File: exercises/domain-2/block-2.2/starter.py
import asyncio
import random
from dataclasses import dataclass, field
from typing import TypeVar, Generic

T = TypeVar("T")


@dataclass
class ToolResult:
    success: bool
    data: dict | None = None
    error: str | None = None
    retryable: bool | None = None
    suggestion: str | None = None


@dataclass
class FetchResult:
    status: int
    body: str


async def simulate_fetch(url: str) -> FetchResult:
    """Simulate HTTP fetch with random failures"""
    rand = random.random()
    if rand < 0.3:
        raise ConnectionError("ECONNREFUSED: connection refused")
    if rand < 0.5:
        return FetchResult(503, "Service Unavailable")
    if "missing" in url:
        return FetchResult(404, "Not Found")
    return FetchResult(200, f"<html>Content of {url}</html>")


async def robust_fetch_tool(url: str) -> ToolResult:
    """
    TODO: Implement robust fetch with retry logic
    - Retry up to 3 times for 5xx / network errors (exponential backoff)
    - Do NOT retry for 4xx errors
    - Return ToolResult (never raise to caller)
    """
    return ToolResult(success=False, error="Not implemented")


async def main():
    print("Testing with valid URL:")
    print(await robust_fetch_tool("https://example.com/page"))

    print("\nTesting with missing URL (404 — should not retry):")
    print(await robust_fetch_tool("https://example.com/missing"))


asyncio.run(main())

File: exercises/domain-2/block-2.2/starter.ts
interface ToolResult<T = unknown> {
  success: boolean
  data?: T
  error?: string
  retryable?: boolean
  suggestion?: string
}

interface FetchResult {
  status: number
  body: string
}

// Simulate an HTTP fetch (replace with real fetch in production)
async function simulateFetch(url: string): Promise<FetchResult> {
  // Randomly fail to simulate real network behavior
  const rand = Math.random()
  if (rand < 0.3) throw new Error("ECONNREFUSED: connection refused")  // network error
  if (rand < 0.5) return { status: 503, body: "Service Unavailable" }  // transient
  if (url.includes("missing")) return { status: 404, body: "Not Found" }  // permanent
  return { status: 200, body: `<html>Content of ${url}</html>` }
}

// TODO: Implement robustFetchTool
// - Retry up to 3 times for 5xx / network errors (exponential backoff)
// - Do NOT retry for 4xx errors
// - Return ToolResult (never throw)
async function robustFetchTool(url: string): Promise<ToolResult<FetchResult>> {
  // TODO: implement
  return { success: false, error: "Not implemented" }
}

// Test it
async function main() {
  console.log("Testing with valid URL:")
  console.log(await robustFetchTool("https://example.com/page"))

  console.log("\nTesting with missing URL (404 — should not retry):")
  console.log(await robustFetchTool("https://example.com/missing"))
}

main()
